lib/Serialization/Deserialization.cpp

```diff
@@ -3344,12 +3346,51 @@ ModuleFile::getDeclCheckedImpl(DeclID DID) {
       error();
       return nullptr;
     }
+    
+    if (opaqueResultTypeDeclID)
+      fn->setOpaqueResultTypeDecl(
+                         cast<OpaqueTypeDecl>(getDecl(opaqueResultTypeDeclID)));
 
     // Set the interface type.
     fn->computeType();
 
     break;
   }
+      
+  case decls_block::OPAQUE_TYPE_DECL: {
+    DeclID namingDeclID;
+    DeclContextID contextID;
+    GenericSignatureID interfaceSigID;
+    TypeID interfaceTypeID;
+    GenericEnvironmentID genericEnvID;
+    SubstitutionMapID underlyingTypeID;
+    
+    decls_block::OpaqueTypeLayout::readRecord(scratch, contextID,
+                                              namingDeclID, interfaceSigID,
+                                              interfaceTypeID, genericEnvID,
+                                              underlyingTypeID);
+    
+    auto opaqueDecl =
+      new (ctx) OpaqueTypeDecl(cast<ValueDecl>(getDecl(namingDeclID)), nullptr,
+                     getDeclContext(contextID),
+                     getGenericSignature(interfaceSigID),
+                     getType(interfaceTypeID)->castTo<GenericTypeParamType>());
+    auto genericEnv = getGenericEnvironment(genericEnvID);
+    opaqueDecl->setGenericEnvironment(genericEnv);
```

Comment:
For later: there's a mechanism to lazily deserialize generic environments. We do this already for generic functions and types; you might want to look into doing it here